CAIP-25: Alternative Schema (with two examples) (#16)
pedrouid/CAIPs: CAIPs/caip-25.md

#### Request

```typescript
interface CAIP25JsonRpcRequest {
  id: number;
  jsonrpc: "2.0";
  method: "wallet_createSession";
  params: {
    scopes: {
      [scopeKey: string]: {
        chains?: string[];
        methods: string[];
        notifications: string[];

>       };

Reviewer: i think it would be a good idea to include the original `accounts` property here where it is used as a suggestion to the wallet for what accounts to have selected by default

Reviewer 2: Agreed w/ Jiexi ☝️ we use this property in our current implemenation

Reviewer 3: my preference is still
mandatory `accounts` array but can be empty > optional `accounts` array > no accounts array allowed in request

Reviewer: let's make this optional array

  result: {
    scopes: {
      [scopeKey: string]: {
        chains?: string[];
        accounts: string[];
        methods: string[];
        notifications: string[];
        capabilities?: {
          [capabilityKey: string]: any;
        };
      };
    };
    properties?: {
      [propertyKey: string]: any;
    };
  };
}
```

Each entry within `scopes` object MAY contain `accounts` and `capabilities` as part of its object for success response.

#### Error Codes

The wallet MAY return generic or specific error messages depending on trust. Trusted responses may include codes like:

- `5000`: Unknown error
- `5001`: User disapproved requested methods
- `5002`: User disapproved requested notifications
- `5100-5102`: Unsupported chains, methods, or notifications
- `5201-5302`: Malformed requests

## Examples

**Example 1**

For request, we define a very simple scope for 10 EVM chains with the exact same scope.

```jsonc
// JSON-RPC REQUEST
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "wallet_createSession",
  "params": {
    "scopes": {
      "wallet": {
        "methods": [
          "wallet_revokeSession",
          "wallet_getSession",
          "wallet_authenticate",

>           "wallet_pay"

Reviewer: we don't want the plain `wallet` scope anymore?

Reviewer: maybe these show up under sessionProperties instead and are non requestable

      "eip155": {
        "chains": [
          "1",
          "10",
          "130",
          "324",
          "2741",
          "8453",
          "42161",
          "59144",
          "534352",
          "747474"
        ],
        "notifications": ["wallet_sessionChanged"]
      },

>       "wallet:eip155": {

Reviewer: sorry. what was our decision around a scope for parking non-chain-specific methods? Are we just saying that they must be parked inside a chain specific scope?

Reviewer 2: I'm personally still on the fence on whether or not we can/should remove this

Reviewer: - We’re replacing wallet:eip155 with eip155:0
- In the eip155 namespace scope, 0 is implicitly included in the chains array
- (^ this means eip155:0 can be directly requested as well)
- In CAIP-27, eip155 is shorthand and gets expanded into eip155:0 (just like how chains array is shorthand and gets expanded)

>         "notifications": ["accountsChanged", "chainChanged"]

Reviewer 2: ```suggestion
        "notifications": ["accountsChanged"]
```

Why would `chainChanged` exist in a CAIP session?

Reviewer: yeah, neither of these should technically

Reviewer 3: on today's call, Al mentioned this should probably not be in the example since it's more legacy/deprecated than normative

```

> For response, we also keep it quite simple with no wallet capabilities also and responded back with just a single account matching all requested EVM chains.

Reviewer 2: I'm not clear on what this comment is adding. I think the "responded back with just a single account matching all requested EVM chains." may be misleading since it seems to suggest you **can only repsond with one account** rather than that this is just a possible response.

}
```

> For the response, we match the same scopes as the request but segregate 2 out of 5 EVM chains into their scopes because of either exclusive accounts or capabilities.

Reviewer 2: ```suggestion
For the response, we match the same scopes as the request but separate 2 out of 5 EVM chains into individual scopes because of exclusive accounts, capabilities or methods.
```

Reviewer 3: "exclusive" might be confusing (sounds like it's about what each chain is capable of, rather than what a particular wallet and user have authorized); "particular" or "bespoke" might be less ambiguous

Reviewer: non-overlapping?

Additionaly we have the two Solana chains returning the same scopes but returning two different account addresses for each chain including a unique capability for one of the chains

Finally the wallet has provided within properties with its walletInfo per [CAIP-372][].

```jsonc
// JSON-RPC RESPONSE
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "sessionId": "0xdeadbeef",
    "scopes": {
      "eip155": {
        "chains": ["1", "10", "324"],
        "accounts": ["0xab16a96d359ec26a11e2c2b3d8f8b8942d5bfcdb"],
        "methods": [
          "eth_sendTransaction",
          "personal_sign",

>           "wallet_grantPermissions",

Author: ```suggestion
          "wallet_addEthereumChain",
          "wallet_grantPermissions",
```
